engine/consensus/matching/engine.go

```diff
@@ -729,10 +675,11 @@ func (e *Engine) ensureStakedNodeWithRole(nodeID flow.Identifier, block *flow.He
 // seals mempool.
 func (e *Engine) sealResult(incorporatedResult *flow.IncorporatedResult) error {
 
-	// store the result to make it persistent for later checks
-	err := e.resultsDB.Store(incorporatedResult.Result)
-	if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
-		return fmt.Errorf("could not store sealing result: %w", err)
+	// at this point we expect to have receipt stored in database. Check our state before sealing.
+	// Failing here means implementation flaw.
+	_, err := e.receiptsDB.ByBlockID(incorporatedResult.Result.BlockID)
+	if err != nil {
+		return fmt.Errorf("missing receipt for execution result: %w", err)
 	}
```

Comment:
I've just added this to make sure that we aren't breaking an invariant.

Reply:
Agree with @arrivets. I also don't think this check necessary. If we don't trust the `storage.receipts.Store(receipt)` call would also store the result, then we could have checked that right after we store the receipt (which I also don't think necessary), instead of here. 

So let's keep it simple, and just trust the storage.receipts does the right job.